### PR TITLE
[crypto] Add AIP-80 input types for private keys

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Adds support for AIP-80 private keys

### Test Plan
See unit tests

### Related Links
Discussion https://github.com/aptos-foundation/AIPs/issues/405

### CODE REVIEW

File: src/core/crypto/ed25519.ts [hunk 1]
```diff
@@ -251,8 +264,8 @@ export class Ed25519PrivateKey extends Serializable implements PrivateKey {
    *
    * @returns string representation of the private key
    */
-  toString(): string {
-    return this.signingKey.toString();
+  toString(): `ed25519-priv-${string}` {
```
Comment (0xmaayan) [inline src/core/crypto/ed25519.ts:267]: isnt it depends on how you generate the privateKey?

File: src/core/crypto/ed25519.ts [hunk 2]
```diff
@@ -147,17 +147,30 @@ export class Ed25519PrivateKey extends Serializable implements PrivateKey {
...
     super();
-
-    const privateKeyHex = Hex.fromHexInput(hexInput);
-    if (privateKeyHex.toUint8Array().length !== Ed25519PrivateKey.LENGTH) {
+    let data: Hex;
+    if (typeof hexInput === "string") {
+      // A non-prefixed input
+      if (!strict && hexInput.startsWith("0x")) {
+        data = Hex.fromHexString(hexInput);
+      } else if (hexInput.startsWith("ed25519-priv-")) {
```
Comment (0xmaayan) [inline src/core/crypto/ed25519.ts:159]: lets have `ed25519-priv-` a global const

File: src/core/crypto/secp256k1.ts [hunk 3]
```diff
@@ -107,16 +107,29 @@ export class Secp256k1PrivateKey extends Serializable implements PrivateKey {
...
     super();
-
-    const privateKeyHex = Hex.fromHexInput(hexInput);
-    if (privateKeyHex.toUint8Array().length !== Secp256k1PrivateKey.LENGTH) {
+    let data: Hex;
+    if (typeof hexInput === "string") {
+      // A non-prefixed input
+      if (!strict && hexInput.startsWith("0x")) {
+        data = Hex.fromHexInput(hexInput);
+      } else if (hexInput.startsWith("secp256k1-priv-")) {
```
Comment (0xmaayan) [inline src/core/crypto/secp256k1.ts:119]: same. global const.

File: src/core/crypto/secp256k1.ts [hunk 4]
```diff
@@ -207,8 +220,8 @@ export class Secp256k1PrivateKey extends Serializable implements PrivateKey {
    *
    * @returns string representation of the private key
    */
-  toString(): string {
-    return this.key.toString();
+  toString(): `secp256k1-priv-${string}` {
+    return `secp256k1-priv-${this.key.toString()}`;
```
Comment (0xmaayan) [inline src/core/crypto/secp256k1.ts:224]: same question as in ed

File: src/core/crypto/ed25519.ts [hunk 5]
```diff
@@ -147,17 +147,30 @@ export class Ed25519PrivateKey extends Serializable implements PrivateKey {
...
-  constructor(hexInput: HexInput) {
+  constructor(hexInput: HexInput, strict: boolean = false) {
     super();
-
-    const privateKeyHex = Hex.fromHexInput(hexInput);
-    if (privateKeyHex.toUint8Array().length !== Ed25519PrivateKey.LENGTH) {
+    let data: Hex;
+    if (typeof hexInput === "string") {
+      // A non-prefixed input
+      if (!strict && hexInput.startsWith("0x")) {
```
Comment (0xmaayan) [inline src/core/crypto/ed25519.ts:157]: `fromHexInput` uses `fromHexString` - do we really need this check?

**Review by 0xmaayan — COMMENTED** [review]
how does the account creation look like with aip-80? 
it looks like, with this change, there is no way for you to create an account with an aip-80 compatible private key
```
const account = Account.generate() // always produces a random private key with no prefix
```

**Comment by gregnazario** [comment]
> how does the account creation look like with aip-80? it looks like, with this change, there is no way for you to create an account with an aip-80 compatible private key
> 
> ```
> const account = Account.generate() // always produces a random private key with no prefix
> ```

The private key is the same, it's just the string representation is different with a prefix to differentiate between other similarly hex encoded inputs

File: src/core/crypto/ed25519.ts (hunk 5 above)
Comment (gregnazario) [reply src/core/crypto/ed25519.ts:157]: Need it to tell that it's not prefixed

**Comment by 0xmaayan** [comment]
> > how does the account creation look like with aip-80? it looks like, with this change, there is no way for you to create an account with an aip-80 compatible private key
> > ```
> > const account = Account.generate() // always produces a random private key with no prefix
> > ```
> 
> The private key is the same, it's just the string representation is different with a prefix to differentiate between other similarly hex encoded inputs

So the real change here is how we represent the private key as a string (with a prefix) and then adding support to when users pass this string into the sdk

**Comment by 0xmaayan** [comment]
@gregnazario what is the progress with that?

**Pull request closed without merging**